Data/ByteString.hs

```diff
@@ -1078,6 +1080,20 @@ index ps n
     | otherwise      = ps `unsafeIndex` n
 {-# INLINE index #-}
 
+-- | 'ByteString' index that returns 'Nothing' if:
+--      @n < 0@
+--      @n > length byteString@
+indexMaybe :: ByteString -> Int -> Maybe Word8
+indexMaybe ps n
+    | n < 0          = Nothing
+    | n >= length ps = Nothing
+    | otherwise      = Just $ ps `unsafeIndex` n
+{-# INLINE indexMaybe #-}
+
+-- | Infix equivalent for `indexMaybe`
+(!?) :: ByteString -> Int -> Maybe Word8
```

Comment:
To be honest, I don't think we should add `!?` just yet -- it's merely syntax sugar for using `indexMaybe` in infix position, and risks stealing syntax from the precious operator namespace. Just adding `indexMaybe` seems less controversial at this point to me.

Reply:
Since I don't see a lot of demand for these additions, I tend to agree with this view. It's not very important to me, though.

Reply:
> Since I don't see a lot of demand for these additions, I tend to agree with this view.

https://github.com/haskell/bytestring/issues/231 changes the picture for me. I'm fine with including `(!?)`.

Reply:
I'm fine either way but I'm slightly in favour of including `(!?)`.